galerautils/src/gu_enc_mmap.cpp

```diff
@@ -0,0 +1,628 @@
+    unsigned long long page_start = (addr_u / pageSize_) * pageSize_;
+    return reinterpret_cast<unsigned char*>(page_start);
+}
+
+size_t EncMMap::page_number(unsigned char* addr) const {
+    unsigned long long offset = ptr2ull(addr) - ptr2ull(base_);
+    size_t pageNo = offset / pageSize_;
+    return pageNo;
+}
+
+
+// encrption / decryption
+void EncMMap::encrypt(unsigned char* dst, unsigned char* src, size_t size, size_t pageNumber) const {
+    // the last page may be not full
+    size = IS_LAST_PAGE(pageNumber) ? lastPageSize_ : size;
+#if REAL_ENCRYPTION
+    size_t pageStartOffset = pageNumber * pageSize_;
+    size_t unencryptedSize = 0;
+
+    if (gu_unlikely(pageStartOffset < encryptionStartOffset_)) {
+        unencryptedSize = std::min(size, encryptionStartOffset_);
+        memcpy(dst, src, unencryptedSize);
+        dst += unencryptedSize;
+        src += unencryptedSize;
+    }
+
+    int encryptedSize = size - unencryptedSize;
+    if(encryptedSize > 0) {
+      encryptor_.set_stream_offset(pageStartOffset + unencryptedSize);
+      encryptor_.encrypt(dst, src, size - unencryptedSize);
+    }
+#else
+    memcpy(dst, src, size);
+#endif
+}
+
+void EncMMap::decrypt(unsigned char* dst, unsigned char* src, size_t size, size_t pageNumber) const {
+    // the last page may be not full
+    size = IS_LAST_PAGE(pageNumber) ? lastPageSize_ : size;
+#if REAL_ENCRYPTION
+    size_t pageStartOffset = pageNumber * pageSize_;
+    size_t unencryptedSize = 0;
+
+    if (gu_unlikely(pageStartOffset < encryptionStartOffset_)) {
+        unencryptedSize = std::min(size, encryptionStartOffset_);
+        memcpy(dst, src, unencryptedSize);
+        dst += unencryptedSize;
+        src += unencryptedSize;
+    }
+
+    int encryptedSize = size - unencryptedSize;
+    if(encryptedSize > 0) {
+        decryptor_.set_stream_offset(pageStartOffset + unencryptedSize);
+        decryptor_.decrypt(dst, src, size - unencryptedSize);
+    }
+#else
+    memcpy(dst, src, size);
+#endif
+}
+
+
+// signal handler install and dispatch
+static std::once_flag signal_handler_once;
+static struct sigaction oldsigact;
+
+void signal_handler(int sig, siginfo_t* info, void* ctx) {
+    // If someone is accessing encMMaps, just bail out without even trying
+    // to find our object.
+    // It can happen in the following situations:
+    // 1. (rare) The client in registering/deregistering new EncMMap object
+    // 2. (more probable but still not so often) Two signal handlers are called
+    //    simultaneously (2 threads). In such a case the 2nd one will retry
+    //    in a while.
+    if (encMmapsLock.test_and_set(std::memory_order_acquire)) {
+        S_DEBUG("signal_handler collision\n");
+        return;
+    }
+    unsigned char *addr = static_cast<unsigned char*>(info->si_addr);
+    S_DEBUG("addr: x%llX\n", ptr2ull(addr));
+    EncMMap*  encmmap = getEncMMap(addr);
+    encMmapsLock.clear(std::memory_order_release);
```

Comment:
Is it safe to clear this here? In theory, other threads could delete encmmap after this, since the signal handler isn't holding a shared_ptr.

Reply:
We are clearing it here to allow other threads which use other EncMMap objects to work in parallel. So this only protects encMMaps container from new objects additions/deletions which can happen in parallel.
EncMMap object related to the current signal cannot be deleted while the signal is active because that would mean that the application is destroying EncMMap and at the same time is using it.